### PR TITLE
STAT code, fixes & review comments by Serge.

### PR DESCRIPTION
**Pull request by anjalidk**

### CODE REVIEW

File: tests/STAT/stat_test.c [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+				all_cpu->guest_nice);
+		all_cpu++;
+	}
+	fprintf(fp, "intr %s\n", st->intr);
+	fprintf(fp, "ctxt %Lu\n",st->ctxt);
+	fprintf(fp, "btime %Lu\n", st->btime);
+	fprintf(fp, "processes %Lu\n", st->processes);
+	fprintf(fp, "procs_running %Lu\n", st->procs_running);
+	fprintf(fp, "procs_blocked %Lu\n", st->procs_blocked);
+	fprintf(fp, "softirq %s\n", st->softirq);
```
Comment (hallyn) [inline tests/STAT/stat_test.c:75]: I don't see anything in the path of calling res_read() that ensures that st->softirq won't be either NULL or garbage memory.  populate_statinfo() should make sure that if no softirq section is found, it intializes the softirq to something safe.

**Comment by anjalidk** [comment]
> On Nov 1, 2023, at 7:59 PM, Serge Hallyn ***@***.***> wrote:
> 
> 
> @hallyn commented on this pull request.
> In tests/STAT/stat_test.c:
> > + all_cpu->idle,
> + all_cpu->iowait,
> + all_cpu->irq,
> + all_cpu->softirq,
> + all_cpu->steal,
> + all_cpu->guest,
> + all_cpu->guest_nice);
> + all_cpu++;
> + }
> + fprintf(fp, "intr %s\n", st->intr);
> + fprintf(fp, "ctxt %Lu\n",st->ctxt);
> + fprintf(fp, "btime %Lu\n", st->btime);
> + fprintf(fp, "processes %Lu\n", st->processes);
> + fprintf(fp, "procs_running %Lu\n", st->procs_running);
> + fprintf(fp, "procs_blocked %Lu\n", st->procs_blocked);
> + fprintf(fp, "softirq %s\n", st->softirq);
> 
> I don't see anything in the path of calling res_read() that ensures that st->softirq won't be either NULL or garbage memory. populate_statinfo() should make sure that if no softirq section is found, it intializes the softirq to something safe.

Yes, thank you, done and pushed above change into rev3.

**Comment by anjalidk** [comment]
Hi Serge,

Any update on this? I submitted the changes you asked for below.
Anjali

File: tests/STAT/stat_test.c [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	struct cpu_stat *all_cpu;
+
+	cpu_num = sysconf(_SC_NPROCESSORS_ONLN);
+	size = sizeof(struct cpu_stat) * cpu_num;
+	stats.all_cpu = (struct cpu_stat *) malloc(size);
+	err = res_read(RES_STAT_INFO, &stats, sizeof(stats)+size, NULL, 0, 0);
+	if (err != 0)
+		printf("err is %d\n",err);
+	st = &stats;
+	fp = fopen ("./stat_info.txt", "w");
```
Comment (hallyn) [inline tests/STAT/stat_test.c:40]: You're not checking fopen() return value.  I know it's just a test, but it just looks bad :(

File: stat.c [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+				&st->cpu.guest_nice);
+	}
+	cstr = strstr(buffer, "intr ");
+	memset(&st->intr, '\0', sizeof(st->intr));
+	if (cstr) {
+		unsigned int bytes = 0;
+		char *cstr1 = cstr + 5;
+		while (*cstr1 != '\n') {
+			cstr1++;
+			bytes++;
```
Comment (hallyn) [inline stat.c:92]: Should probably check bytes against max size st->intr can support.

File: stat.c [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	if (cstr)
+		sscanf(cstr, "procs_blocked %Lu", &st->procs_blocked);
+	cstr = strstr(buffer, "softirq ");
+	memset(&st->softirq, '\0', sizeof(st->softirq));
+	if (cstr) {
+		unsigned int bytes = 0;
+		char *cstr1 = cstr + 8;
+		while (*cstr1 != '\n') {
+			cstr1++;
+			bytes++;
```
Comment (hallyn) [inline stat.c:118]: again, check bytes length.

**Comment by anjalidk** [comment]
> On Dec 5, 2023, at 11:48 AM, Serge Hallyn ***@***.***> wrote:
> 
> 
> @hallyn commented on this pull request.
> In stat.c:
> > + if (cstr)
> + sscanf(cstr, "processes %Lu", &st->processes);
> + cstr = strstr(buffer, "procs_running ");
> + if (cstr)
> + sscanf(cstr, "procs_running %Lu", &st->procs_running);
> + cstr = strstr(buffer, "procs_blocked ");
> + if (cstr)
> + sscanf(cstr, "procs_blocked %Lu", &st->procs_blocked);
> + cstr = strstr(buffer, "softirq ");
> + memset(&st->softirq, '\0', sizeof(st->softirq));
> + if (cstr) {
> + unsigned int bytes = 0;
> + char *cstr1 = cstr + 8;
> + while (*cstr1 != '\n') {
> + cstr1++;
> + bytes++;
> 
> again, check bytes length.

Thanks, taken care of all 3 comments and did push to rev3.

Anjali

**Comment by anjalidk** [comment]
Hi Serge, any update on this?

**Review by hallyn — APPROVED** [review]

**Comment by hallyn** [comment]
It looks fine to me - but tests are failling?

**Comment by anjalidk** [comment]
Yes, could you merge this review? Part of the fix is in this commit (for stat) and cpu I will submit as soon as you merge this one.

Anjali

**Comment by hallyn** [comment]
Sure - but I approved it, does that not give you the ability to merge it?

**Comment by anjalidk** [comment]
Ok, I did not know you expected me to merge it. Will do so next time.

Anjali

**Comment by hallyn** [comment]
Didnt expect, but it as intended to be an option :)

**Pull request merged**